### ISSUE TITLE
Add a new struct to replace ArrayOrString struct

### ISSUE THREAD

**Issue by ywluogg** [tektoncd/pipeline#4858]
With implementation of [TEP 075 and 076](https://github.com/tektoncd/pipeline/issues/4723) going on, especially introducing json objects into params and results, it worths discussion about whether we want a new struct to replace [ArrayOrString](https://github.com/tektoncd/pipeline/blob/1c8447d74285cbc0b6947822ca88b9d27829c0ba/pkg/apis/pipeline/v1beta1/param_types.go#L134), as the implementation of json objects will happen in ArrayOrString, but the name doesn't really reflect what it actually represents. With [this discussion](https://github.com/tektoncd/pipeline/pull/4786#discussion_r862023838), we are exploring two options here to reduce library breakages, and have a better structure.
- replace ArrayOrString with brand new struct
  - pros:
    - Have the potential to redesign the structure of the struct, so that we can cover some other flexibilities outside of TEP 075 and 076
  - cons:
    - having a tailing old struct that would be similar to the new struct
- rename ArrayOrString
  - pros:
    - don't need to think about a new structure, and the current structure is still flexible, with the scope of TEP 075 and 076
  - cons:
    - current structure could be not flexible for some of TEPs that's not [075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md) and [076](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md)
@vdemeester @pritidesai I'm trying follow up on the discussion we have about this topic in WG. I think based on the above pros and cons comparisons, I'm still leaning toward changing the name of the struct, rather than introducing a new struct.

Pinging @chuangw6 @Yongxuanzhang as well.

**Comment by vdemeester** [comment]
@ywluogg I think, in the end, it's all an implementation detail. What we can do, if we want to "take" our time and split the work into multiple PRs, is to create the new struct as we see it fit, write tests for it. And then switch ArrayOrString with it, and finally remove ArrayOrString. This would be the best of both world, creating the new struct but with the intent and plan of replacing `ArrayOrString`.

**Comment by pritidesai** [comment]
We can have a list of PRs. Here is what I am thinking:

* Create new structs along with tests - `ParamTypes` and `ResultTypes` both can inline the same struct `TypeSpec` which has `string`, `array`,  and  `map` (please feel free to rename if you can find better names).

* Switch params and results with this new type and at the same time, dashboard and CLI can switch using this new type. CLI heavily relies on the existing type:

![Screen Shot 2022-05-19 at 10 55 55 AM](https://user-images.githubusercontent.com/206285/169366107-c18a79c8-3ba4-4f25-8376-799124deb6c7.png)

* Remove legacy `ArrayOrString`

**Comment by pritidesai** [comment]
I had a conversation with @vinamra28 about this plan. It looks great except the last PR - removing legacy `ArrayOrString`. Since CLI and other projects relies on it heavily, we will maintain the legacy `struct` until we have approval from the dependent projects i.e. the dependent projects have transitioned away from this legacy `struct`.

**Comment by chuangw6** [comment]
> I had a conversation with @vinamra28 about this plan. It looks great except the last PR - removing legacy `ArrayOrString`. Since CLI and other projects relies on it heavily, we will maintain the legacy `struct` until we have approval from the dependent projects i.e. the dependent projects have transitioned away from this legacy `struct`.

Thanks @pritidesai for bringing up this. We should definitely take the dependent projects into consideration. Not sure how long it would take for the dependent projects to transition away from this legacy struct.

If it'd happen not too late, how about we don't change the current version of [`ArrayOrString`](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md#emitting-object-results) that has already included the `objectVal` (no rename, no restruct)? Just keep it as is though the name doesn't quite reflect the inclusion of object for now. Once we have approval from the dependent projects, we can just rename it. Because the [TEP-75 section](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md#emitting-object-results) mentioned that  
> We would use the same [ArrayOrString type](https://github.com/tektoncd/pipeline/blob/1f5980f8c8a05b106687cfa3e5b3193c213cb66e/pkg/apis/pipeline/v1beta1/param_types.go#L89) (**expanded to support dictionaries**, and in TEP-0074, arrays) for task results, e.g. for the above example, the TaskRun would contain:

**Comment by chuangw6** [comment]
> I had a conversation with @vinamra28 about this plan. It looks great except the last PR - removing legacy `ArrayOrString`. Since CLI and other projects relies on it heavily, we will maintain the legacy `struct` until we have approval from the dependent projects i.e. the dependent projects have transitioned away from this legacy `struct`.

Thanks @pritidesai for bringing up this. We should definitely take the dependent projects into consideration. So to help the dependent projects to transition away from this legacy struct, a new struct will be needed while the legacy one is maintained. Is that right?

**Comment by vinamra28** [comment]
> > I had a conversation with @vinamra28 about this plan. It looks great except the last PR - removing legacy `ArrayOrString`. Since CLI and other projects relies on it heavily, we will maintain the legacy `struct` until we have approval from the dependent projects i.e. the dependent projects have transitioned away from this legacy `struct`.
> 
> Thanks @pritidesai for bringing up this. We should definitely take the dependent projects into consideration. So to help the dependent projects to transition away from this legacy struct, a new struct will be needed while the legacy one is maintained. Is that right?

@chuangw6 yes the legacy one needs to be maintained for some time along with the new one. Also dependent projects(atleast CLI) can start their migration once this is out as CLI project vendors the pipelines project.

**Comment by chuangw6** [comment]
Thanks @vinamra28!

If a new struct is created that embeds `ArrayOrString` and includes object type, the type for `ParamSpec`'s [`Default`](https://github.com/tektoncd/pipeline/blob/7aef957effb55a519bba97c9d2c36a28482b35d4/pkg/apis/pipeline/v1beta1/param_types.go#L55) field and Param's [`value`](https://github.com/tektoncd/pipeline/blob/7aef957effb55a519bba97c9d2c36a28482b35d4/pkg/apis/pipeline/v1beta1/param_types.go#L110) field will be changed to the new struct. Would this affect the dependent projects?

**Comment by vdemeester** [comment]
Chiming here a tad:

if the shape of the api doesn't change aka:
- if the way we marshal to yaml for the same type doesn't change)
- if the name of the fields do not change
… then it's not a problem for the CLI.

It's similar to "duck typing", if it quacks it's a duck. The go struct and internal representation of our type can change without breaking the api, and thus without breaking older clients. My assumption is that we would have that new struct be compatible with `ArrayOrString` (at least in how the current behavior of `ArrayOrString` is). This would make a switch of struct transparent for today's behavior. When we add additional "feature" to that new struct, then old client will work with old behavior but not new one, which is to be expected.

*Note: the cli could be defining it's own set of struct representing the API (it would be even more true if the cli was written in another language). From `pipeline` what we need to be careful about is the API shape, not necessarily our go code and struct*.

**Comment by ywluogg** [comment]
> My assumption is that we would have that new struct be compatible with `ArrayOrString` (at least in how the current behavior of `ArrayOrString` is).

Combining all are considered above, let's have a more defined new struct here to make sure we understand the compatibility in the same way.
The new struct should have the following:
```
type StructuredValue struct { 
        // Old fields
	Type      StructuredValueType `json:"type"` // Represents the stored type of ArrayOrString.
	StringVal string    `json:"stringVal"`
	// +listType=atomic
	ArrayVal  []string          `json:"arrayVal"`
	ObjectVal map[string]string `json:"objectVal"`
        // New fields that only work for new feature
        NewField NewfieldType
}
```

**Comment by ywluogg** [comment]
Update on this issue: synced offline with @pritidesai @vdemeester that we agree that the changes wouldn't break any of the existing APIs but will introduce breakages in library. We will go with renaming the old struct, as we already added objectType in ArrayOrString@

**Comment by lbernick** [comment]
Closed by #5304
